resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageManager.scala

```diff
@@ -97,11 +96,54 @@ private[spark] class SparkDockerImageBuilder
     buildImage("spark-integration-test-asset-server", STATIC_ASSET_SERVER_DOCKER_FILE)
   }
 
+  def deleteImages(): Unit = {
+    deleteImage("spark-driver")
+    deleteImage("spark-driver-py")
+    deleteImage("spark-executor")
+    deleteImage("spark-executor-py")
+    deleteImage("spark-shuffle")
+    deleteImage("spark-resource-staging-server")
+    deleteImage("spark-init")
+    deleteImage("spark-integration-test-asset-server")
+    deleteImage("spark-base")
+  }
+
   private def buildImage(name: String, dockerFile: String): Unit = {
-    dockerClient.build(
-      DOCKER_BUILD_PATH,
-      name,
-      dockerFile,
-      new LoggingBuildHandler())
+    log.info(s"Building Docker image - $name:$dockerTag")
+    val dockerFileWithBaseTag = new File(DOCKER_BUILD_PATH.resolve(
+        s"$dockerFile-$dockerTag").toAbsolutePath.toString)
+    dockerFileWithBaseTag.deleteOnExit()
+    try {
+      val originalDockerFileText = Files.readLines(
+          DOCKER_BUILD_PATH.resolve(dockerFile).toFile, Charsets.UTF_8).asScala
+      val dockerFileTextWithProperBaseImage = originalDockerFileText.map(
+          _.replace("FROM spark-base", s"FROM spark-base:$dockerTag"))
+      Utils.tryWithResource(Files.newWriter(dockerFileWithBaseTag, Charsets.UTF_8)) { fileWriter =>
+        Utils.tryWithResource(new PrintWriter(fileWriter)) { printWriter =>
+          for (line <- dockerFileTextWithProperBaseImage) {
+            // scalastyle:off println
+            printWriter.println(line)
+            // scalastyle:on println
+          }
+        }
+      }
+      dockerClient.build(
+          DOCKER_BUILD_PATH,
+          s"$name:$dockerTag",
+          s"$dockerFile-$dockerTag",
+          new LoggingBuildHandler())
+    } finally {
+      dockerFileWithBaseTag.delete()
+    }
+  }
+
+  private def deleteImage(name: String): Unit = {
+    try {
+      dockerClient.removeImage(s"$name:$dockerTag")
```

Comment:
If I understand this right, you're modifying the system global docker images here. Given that dockerTag is always 'latest' is there a concern that multiple simultaneous invocations of this test could be fighting for the same 'latest' image? Should we change the tag to have a unique value per build?

Reply:
I see you just answered this in the other comment thread.

Reply:
The tag is actually generated here in the Minikube backend.

Reply:
Tag should be set to a UUID. I also considered the git hash as mentioned before.